> I think we should rewrite:
> 
> ```
> <Unknown>:<Unknown>: note: Non-deterministic values: __VERIFIER_nondet_int:test-0002.c:9:9: len 4 bytes, [4 times 0x0] (i32: 0)
> ```
> 
> like this:
> 
> ```
> test-0002.c:9:9: note: Non-deterministic values: __VERIFIER_nondet_int: len 4 bytes, [4 times 0x0] (i32: 0)
> ```
> 
> That is the format that gcc-compatible compilers use by default (and what csdiff recognizes).

Ok, I'll rewrite it like that.






> I've just found another problem. All lines from call stack repeat the location info twice:
> 
> ```
> /home/lukas/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10: note: call stack: function main () at: /home/luka
> s/aufover-benchmark/workdir/single-c/mem-basic-realloc/0029-realloc-invalid-ptr-freed-2/symbiotic/0029-test.c:10
> ```

Not exactly, the first location is based on the location of the error, the second location is based on the function call: 
```
#include <stdlib.h>

int bar ()
{
        int *a = NULL;
        *a = 10;
        return *a;
}

int foo ()
{
        return bar();
}


int main ()
{
        return foo();

}
```

```foo.c:6: error: memory error: out of bound pointer
foo.c:6: note: call stack: function bar () at: foo.c:6
foo.c:6: note: call stack: function foo () at: foo.c:12
foo.c:6: note: call stack: function main () at: foo.c:18
foo.c:6: note: Additional Info: address: 0:0
foo.c:6: note: Additional Info: pointing to: none
```
(using an old version of symbiotic, thus the relative paths)

--- a/symbiotic2cs.py
+++ b/symbiotic2cs.py
@@ -1,9 +1,10 @@
 #!/bin/python3
 import sys
 import re
+import argparse
 
 class Error_trace:
-    def __init__(self,argv = ""):
+    def __init__(self,argv = "", pr_stack = True, pr_info = True, pr_nondet = True):
         self.file = "<Unknown>"
         self.line = "<Unknown>"
         self.summary = "<Unknown>"
@@ -11,9 +12,9 @@ class Error_trace:
         self.info = ""
         self.nondet_values = ""
         self.argv = argv
-        self.print_stack = True
-        self.print_info = True
-        self.print_nondet_values = True
+        self.print_stack = pr_stack
+        self.print_info = pr_info
+        self.print_nondet_values = pr_nondet
 
     def fix_file(self):
         'Check whether the file is a file from the input package or file from some symbiotic library. If is the laterthis function tries to find a proper file name/line number from the stackinfo'
@@ -173,8 +174,22 @@ class Parser:
 
 if __name__ == "__main__":
     symbiotic_return_value = 0
-    input_str = sys.stdin.read()
-    parser = Parser()
-    for l in str.splitlines(input_str):
-        parser.send(l)
+
+    argv_parser = argparse.ArgumentParser(description='Convert the symbiotic output to csgrep format.')
+    argv_parser.add_argument('input_files', metavar='FILE', nargs='*',  help='input file to be converted')
+    args = argv_parser.parse_args()
+
+    #Input is read from the stdin
+    if args.input_files == []:
+        input_str = sys.stdin.read()
+        input_parser = Parser()
+        for l in str.splitlines(input_str):
+            input_parser.send(l)
+    #The input files are part of the argv
+    else:
+        for f in args.input_files:
+            input_str = open(f,"r")
+            input_parser = Parser()
+            for l in input_str:
+                input_parser.send(l)
     sys.exit(symbiotic_return_value)

--- a/symbiotic2cs.py
+++ b/symbiotic2cs.py
@@ -36,23 +36,15 @@ class Error_trace:
         header = ("Error: SYMBIOTIC_WARNING:\n")
         self.fix_file()
         summary = self.file + ":" + self.line + ": error: " + self.summary +"\n"
-        new_stack = ""
         new_info = ""
-        new_nondet_values = ""
-        if self.print_stack:
-            for l in str.splitlines(self.stack):
-                new_stack += self.file + ":" + self.line + ": " + l + "\n"
         if self.print_info:
             for l in str.splitlines(self.info):
                 new_info += self.file + ":" + self.line + ": " + l + "\n"
-        if self.print_nondet_values:
-            for l in str.splitlines(self.nondet_values):
-                new_nondet_values += "<Unknown>" + ":" + "<Unknown>" + ": " + l + "\n"
         if self.argv != "" :
             new_argv = self.file + ":" + self.line + ": note: argv: " + self.argv + "\n"
         else:
             new_argv = ""
-        return header + summary + new_argv + new_stack + new_info + new_nondet_values
+        return header + summary + new_argv + self.stack + new_info + self.nondet_values
 
 class Parser:
     'state transitions:'
@@ -135,7 +127,7 @@ class Parser:
                 'Skip the debug messages'
             elif re.search("\s+(.*)\s+in\s+(.*)\s+at\s*(.*)\s*", token):
                 m = re.search("\s+(.*)\s+in\s+(.*)\s+at\s*(.*)\s*", token)
-                self.current_trace.stack += "note: call stack: function " + m.group(2) + " at: " + m.group(3) + "\n"
+                self.current_trace.stack += m.group(3) + ": " + "note: call stack: function " + m.group(2) + "\n"
 
     def _create_state_info(self):
         while True:
@@ -151,7 +143,8 @@ class Parser:
                 'Skip the debug messages'
             elif re.search("\s+(.*)", token):
                 m = re.search("\s+(.*)", token)
-                self.current_trace.info += "note: Additional Info: " + m.group(1) + "\n"
+                if not re.search("^\s+$",token):
+                    self.current_trace.info += "note: Additional Info: " + m.group(1) + "\n"
 
     def _create_state_nondet_values(self):
         while True:
@@ -169,7 +162,7 @@ class Parser:
                 m = re.search("\s*(.*):(.*):(.*):(.*) :=\s*(.*)\s*", token)
                 '__VERIFIER_nondet_int:test-0002.c:9:9 := len 4 bytes, [4 times 0x0] (i32: 0)'
                 'TODO: map the group(1) to something more user friendly?'
-                self.current_trace.nondet_values += "note: Non-deterministic values: " + m.group(1) + ":" + m.group(2) + ":" + m.group(3) + ":" + m.group(4) + ": " + m.group(5)  + "\n"
+                self.current_trace.nondet_values += m.group(2) + ":" + m.group(3) + ":" + m.group(4) + ": " + "note: Non-deterministic values: " + m.group(1) + ": " + m.group(5)  + "\n"
 
     def _create_state_trap(self):
         while True:

--- a/symbiotic2cs.py
+++ b/symbiotic2cs.py
@@ -96,6 +96,9 @@ class Parser:
             if re.search("--- Error trace ---", token):
                self.current_trace = Error_trace(argv = self.argv)
                self.current_state = self.state_error_trace
+            elif re.search ("RESULT: (ERROR|error)\s*(.*)",token):
+                m = re.search ("RESULT: (ERROR|error)\s*(.*)",token)
+                print ("Error: SYMBIOTIC_WARNING:\n" + "symbiotic: internal error: " + m.group(2) + "\n" + "note: argv: " + self.argv + "\n")
             elif re.search ("\[DBG\] Argv:\s+(.*)",token):
                 self.argv = re.search ("\[DBG\] Argv:\s+(.*)",token)[1]
 

